Option automatically to pause on subtitle end

twevs opened the pull request:
Before anything else, thank you for producing this amazing software.

In trying it, however, I felt it was lacking a few [Language Reactor playback features](https://www.languagereactor.com/help/basic), so I decided to add them for my own comfort, starting with autopause on subtitle end. I have added an "Automatically pause the player at the end of a subtitle" checkbox, unchecked by default, in the "Behavior" settings.

It seems to be a popular enough feature among language learners that I thought it worth cleaning up and opening a PR for.

(As an habitual MPC-HC user, I started writing this before realizing just how extensible through scripts mpv actually is, and subsequently found that there are two scripts, [sub-pause](https://github.com/Ben-Kerman/mpv-sub-scripts) and [sub-replay](https://github.com/kelciour/mpv-scripts/blob/master/sub-replay.lua), which also provide Language Reactor playback controls; however, the former pauses too soon with the .ass files I've tried and the latter only works with .srt files.)

ripose-jp commented on src/gui/widgets/overlay/subtitlewidget.cpp:
```diff
@@ -467,6 +478,14 @@ void SubtitleWidget::adjustVisibility()
+        && !pausedForCurrentSubtitle
+        && (m_subtitle.endTime - value <= 0))
+    {
+         GlobalMediator::getGlobalMediator()->controlsPause();
```
Use `GlobalMediator::getGlobalMediator()->getPlayerAdapter()->pause();` because this isn't sent from PlayerControls.
Also when sending signals be sure to use `Q_EMIT` even though it isn't necessary.

ripose-jp commented on src/gui/widgets/overlay/subtitlewidget.h:
```diff
@@ -212,7 +212,15 @@ private Q_SLOTS:
+
+    /* True if playback has already been paused for the current subtitle, false
+     * otherwise. */
+    bool pausedForCurrentSubtitle = false;
```
This is a member variable so be sure to prefix it with `m_`.
Also move it up below `m_paused` since it isn't a inside a struct.

ripose-jp commented on src/gui/widgets/settings/behaviorsettings.cpp:
```diff
@@ -88,13 +88,19 @@ void BehaviorSettings::restoreSaved()
     );
+    m_ui->checkSubtitlePause->setChecked(
+        settings.value(
+            SETTINGS_BEHAVIOR_SUBTITLE_PAUSE, SETTINGS_BEHAVIOR_SUBTITLE_PAUSE_DEFAULT
```
Keep an 80 character line width.

ripose-jp commented on src/gui/widgets/settings/behaviorsettings.cpp:
```diff
@@ -88,13 +88,19 @@ void BehaviorSettings::restoreSaved()
 {
     m_ui->checkAutofit->setChecked(SETTINGS_BEHAVIOR_AUTOFIT_DEFAULT);
     m_ui->spinAutofit->setValue(SETTINGS_BEHAVIOR_AUTOFIT_PERCENT_DEFAULT);
+    m_ui->checkSubtitlePause->setChecked(SETTINGS_BEHAVIOR_SUBTITLE_PAUSE_DEFAULT);
```
80 characters

ripose-jp commented:
When you fix these things up, be sure to make them as fixup commits to the relevant commits to keep everything clean.
https://stackoverflow.com/questions/3103589/how-can-i-easily-fixup-a-past-commit
You can then do a force push with `git push -f`.
Since you've already made a few commits in response, you can undo them while keeping the changes using `git reset --soft HEAD~<number of commits to undo>` or just keep running `git reset --soft HEAD^` until you make it to the correct commit.

Also consider adding a menu option with a key bind in the menu under Subtitles to toggle this feature as well since going into the settings for something that may be toggled often is annoying. It should go right under "Add Subtitle File". Make sure the key bind doesn't overlap with anything mpv uses if you can help it.

ripose-jp commented:
Also, let me sleep on this one a bit since the feature can be implemented with mpv scripts like you said.
To make me decision a little easier, can you link me to a subtitle file that causes problems with the sub-pause?

ripose-jp commented:
You forgot to do the autosquash step since the `fixup!` commits still show up.
Make sure you run these command before you continue:
`git rebase -i --autosquash <commit id to fixup>~1`

twevs commented:
Thank you very much for your help!

It is getting late here but a key binding is a good idea and I will be sure to add it tomorrow.

[Here](https://kitsunekko.net/subtitles/japanese/Non%20Non%20Biyori/[Tsundere]%20Non%20Non%20Biyori%20[BDRip%20h264%201920x1080%2010bit%20FLAC].rar) are some subtitles that don't work well with sub-pause.

ripose-jp commented:
No worries, take your time to figure out how to fix it.

I'll check out the example subtitles and get back to you.

twevs commented:
Thank you. Git is new to me and it turns out I was on a detached head when (I think) trying to do the autosquashing, and the problems snowballed from there (especially with the fatigue). The commit history has ended up including conflict resolution traces. If you do end up accepting the changes, you can close this PR and I can submit another one from a clean branch, if you'd like. Sorry for the trouble.

ripose-jp commented:
Rather than making a new PR, you can create a new branch from another that's equal with ripose-jp/master, copy all the changes you want from twevs/feat-subpause to that new branch, checkout twevs/feat-subpause, then run `git reset --hard origin/<new branch>`, then force push.

It should look something like this:
```
git checkout master
git checkout -b temp-changes
<copy over all your code from feat-subpause>
<commit everything you want to commit>
git checkout feat-subpause
git reset --hard temp-changes
git push -f
```
When you run `git reset --hard temp-changes` it will completely replace the `feat-subpause` branch with `temp-changes`, so make sure you've backed up all your changes because this will overwrite everything on `feat-subpause`.

ripose-jp commented:
There's a bug. Here's the steps to reproduce:

1. Open Memento without loading a file
2. Make sure auto pause is enabled
3. Load a file
4. The video will start paused

The expected behavior on loading a file is that the player always starts playing.

twevs commented:
Thank you, you're right. Adding
`&& value > DOUBLE_DELAY`
to the conditions triggering a pause in SubtitleWidget::positionChanged seems to be one way of fixing it, on the assumption that a video would never need to pause within that delay. I'd have to check whether that would protect against videos with low frame rates, though, and if not whether it would be reasonable simply to go with a _reasonable_ default threshold.

I should have gone to bed hours ago and I'd rather not make more changes when I'm so tired — thank you very much for your help and patience so far.

ripose-jp commented:
I was testing your PR out a little more and I find it doesn't handle pausing very well if there's another subtitle on the frame after the previous one ends. When it pauses, the next subtitle is almost always on screen. There are a few subtitles at the start of Non Non Biyori episode 8 that do this at the start. You could probably add a larger delta on line 483 of SubtitleWidget so that it pauses sooner. I haven't tested this though.

I've also decided that I'm fine merging this PR as long as it works well.

twevs commented:
Thank you for your feedback.

I have added a menu option with a key binding that doesn't conflict with mpv's to toggle the functionality.

I have also increased the delta on line 483 of SubtitleWidget to 28 ms. (Thankfully I had a branch which had made the delta a menu option in and of itself, which I didn't push because it seemed like overkill but which came in very useful in experimenting here.) I found that 27 ms was the smallest delta that made those subtitles consistently work for me and 29 ms already started eating into the audio; the golden window is therefore very small, but if you'd prefer to err on safety's side we can always bump it up a bit (or, indeed, allow the delta to be specified in the menu).

I included a reasonable initial delay to fix pausing on opening a video and added a condition to protect against initial pausing in case m_subtitle.endTime is initialized to a value less than or equal to PAUSE_DELTA.

ripose-jp commented:
Those should be the last changes you'll need to make unless I find something I missed something. Good work.

ripose-jp reviewed (changes requested):
When I wrote the last comment, I thought I had already requested these changes. My bad.

twevs commented:
Thank you — I have pushed the changes.
